syntax = "proto3";

package NFS;

service nfsServer {
    rpc call (request) returns (response) {
    }
}

message request {
    string name = 1;
    attributes attr = 2;
    int32 todirfh = 3;
    int32 fromfh = 4;
    int32 fh = 5;
    fileInfo fi = 6;
    int32 size = 7;
    int32 offset = 8;
    int32 dirfh = 9;
    string path = 10;
    bytes data = 11;
    int32 isDataSync = 12;
    string command = 13;
    int32 gen = 14;
    repeated int32 gens = 15;
}

message response {
    int32 fh = 1;
    uint64 success = 2;
    uint64 err = 3;
    fileInfo fi = 4;
    bytes data = 5;
    int32 ern = 6;
    int32 size = 7;
    repeated directory objects = 8;
    int32 status = 9;
    attributes attr = 10;
    int32 dataSize = 11;
}

message attributes {
    uint64 st_dev = 1;
    uint64 st_ino = 2;
    uint64 st_mode = 3;
    uint64 st_nlink = 4;
    uint64 st_uid = 5;
    uint64 st_gid = 6;
    uint64 st_rdev = 7;
    uint64 st_size = 8;
    uint64 st_catime = 9;
    uint64 st_cmtime = 10;
    uint64 st_cctime = 11;
    uint64 st_blksize = 12;
    uint64 st_blocks = 13;
    uint64 st_attr = 14;
}


message fileInfo {
    uint64 fh = 1;
    uint32 flags = 2;
}

message directory {
    string name = 1;
    attributes attr = 2;
}

